contractcourt/htlc_timeout_resolver.go

```diff
@@ -450,22 +450,38 @@ func (h *htlcTimeoutResolver) Resolve(
 		return h.claimCleanUp(commitSpend)
 	}
 
-	log.Infof("%T(%v): resolving htlc with incoming fail msg, fully "+
-		"confirmed", h, h.htlcResolution.ClaimOutpoint)
-
 	// At this point, the second-level transaction is sufficiently
 	// confirmed, or a transaction directly spending the output is.
 	// Therefore, we can now send back our clean up message, failing the
-	// HTLC on the incoming link.
-	failureMsg := &lnwire.FailPermanentChannelFailure{}
-	if err := h.DeliverResolutionMsg(ResolutionMsg{
-		SourceChan: h.ShortChanID,
-		HtlcIndex:  h.htlc.HtlcIndex,
-		Failure:    failureMsg,
-	}); err != nil {
+	// HTLC on the incoming link. We also make sure we have not already
+	// failed this HTLC.
+	canceledHTLCs, err := h.FetchCanceledHTLCs()
+	if err != nil {
 		return nil, err
 	}
 
+	if !canceledHTLCs.Contains(h.htlc.HtlcIndex) {
```

Comment:
As mentioned in the db structure, if `cancelHTLCs` has lots of items, it could be slow to create this set.

Reply:
you mean the deserialization? Agree depending on whether the new proposed design is flexible I will change the structure.

Reply:
obsolete comment because db entry is removed.